In [1]:
%load_ext autoreload

In [2]:
%autoreload 2

In [3]:
import os
os.chdir("D:/Projects/insect-id")

import mynnlibv2
from mynnlibv2 import *

In [4]:
dataset_dir = "insect-dataset/lepidoptera"

# Normalize data

In [8]:
import shutil

# make val dataset uniform
min_val_data_cnt = 2
for class_name in os.listdir(f"{dataset_dir}/data"):
    if not os.path.exists(f"{dataset_dir}/val/{class_name}"):
        os.makedirs(f"{dataset_dir}/val/{class_name}")
    val_data_cnt = len(os.listdir(f"{dataset_dir}/val/{class_name}"))
    data_to_add = max(0, min_val_data_cnt - val_data_cnt)
    if data_to_add > 0:
        files = os.listdir(f"{dataset_dir}/data/{class_name}")
        random.shuffle(files)
        for file in files[:data_to_add]:
            shutil.copy2(f"{dataset_dir}/data/{class_name}/{file}", f"{dataset_dir}/val/{class_name}/{file}")

In [14]:
# find inconsistant data
for class_name in os.listdir(f"{dataset_dir}/val"):
    if not os.path.exists(f"{dataset_dir}/data/{class_name}"):
        shutil.rmtree(f"{dataset_dir}/val/{class_name}")
        print(f"[WARNING] class {class_name} in val not present in data removed")

[WARNING] class apona-spp in val not present in data removed
[WARNING] class vindula-dejone in val not present in data removed


In [19]:
validate_dataset(f"{dataset_dir}/data", f"{dataset_dir}/val")

[INFO] train set: 325920 images
[INFO] val set: 11105 images


# Model A

 - RESNET-152
 - lr = 1e-4
 - rubustness (random crop/resize/jitter) increased by 0.1 on each epoch
 - model feature_extractor & fc modified

## Itertaion 1
- 5615 classes with 325920 images

In [5]:
model_data = init_model(f"{dataset_dir}/data", f"{dataset_dir}/val", batch_size=32, image_size=224, lr=1e-4, validate=False)
for epoch in range(15):
    model_data = run_epoch(model_data, output_path=f"{dataset_dir}/checkpoint.inc.lepidoptera.ta", robustness_lambda=0.1)

[INFO] classes: 5615
[INFO] Training started at 2025-06-08 21:21:46


[INFO] Iteration 01 | Epoch 01 | Train      --> 10185/10185 batches | Elapsed time: 02:00:10 | Loss: 2.39 | Acc: 0.573


[INFO] Iteration 01 | Epoch 01 | Val        --> 348/348 batches | Elapsed time: 00:02:39 | Loss: 2.02 | Acc: 0.637


[INFO] Iteration 01 | Epoch 02 | Train      --> 10185/10185 batches | Elapsed time: 01:58:25 | Loss: 0.742 | Acc: 0.811


[INFO] Iteration 01 | Epoch 02 | Val        --> 348/348 batches | Elapsed time: 00:02:45 | Loss: 0.714 | Acc: 0.831


[INFO] Iteration 01 | Epoch 03 | Train      --> 10185/10185 batches | Elapsed time: 01:59:45 | Loss: 0.478 | Acc: 0.864


[INFO] Iteration 01 | Epoch 03 | Val        --> 348/348 batches | Elapsed time: 00:02:45 | Loss: 0.34 | Acc: 0.904


[INFO] Iteration 01 | Epoch 04 | Train      --> 10185/10185 batches | Elapsed time: 01:56:50 | Loss: 0.356 | Acc: 0.891


[INFO] Iteration 01 | Epoch 04 | Val        --> 348/348 batches | Elapsed time: 00:02:43 | Loss: 0.209 | Acc: 0.939


[INFO] Iteration 01 | Epoch 05 | Train      --> 10185/10185 batches | Elapsed time: 02:05:13 | Loss: 1.32 | Acc: 0.704


[INFO] Iteration 01 | Epoch 05 | Val        --> 348/348 batches | Elapsed time: 00:02:41 | Loss: 0.339 | Acc: 0.902


[INFO] Iteration 01 | Epoch 06 | Train      --> 10185/10185 batches | Elapsed time: 02:04:04 | Loss: 1.12 | Acc: 0.745


[INFO] Iteration 01 | Epoch 06 | Val        --> 348/348 batches | Elapsed time: 00:02:47 | Loss: 0.249 | Acc: 0.925


[INFO] Iteration 01 | Epoch 07 | Train      --> 10185/10185 batches | Elapsed time: 02:05:33 | Loss: 1.03 | Acc: 0.764


[INFO] Iteration 01 | Epoch 07 | Val        --> 348/348 batches | Elapsed time: 00:02:48 | Loss: 0.236 | Acc: 0.932


[INFO] Iteration 01 | Epoch 08 | Train      --> 10185/10185 batches | Elapsed time: 02:05:43 | Loss: 0.961 | Acc: 0.778


[INFO] Iteration 01 | Epoch 08 | Val        --> 348/348 batches | Elapsed time: 00:02:49 | Loss: 0.236 | Acc: 0.928


[INFO] Iteration 01 | Epoch 09 | Train      --> 10185/10185 batches | Elapsed time: 02:06:54 | Loss: 0.914 | Acc: 0.788


[INFO] Iteration 01 | Epoch 09 | Val        --> 348/348 batches | Elapsed time: 00:02:50 | Loss: 0.219 | Acc: 0.936


[INFO] Iteration 01 | Epoch 10 | Train      --> 10185/10185 batches | Elapsed time: 02:09:48 | Loss: 0.879 | Acc: 0.796


[INFO] Iteration 01 | Epoch 10 | Val        --> 348/348 batches | Elapsed time: 00:02:53 | Loss: 0.196 | Acc: 0.941


[INFO] Iteration 01 | Epoch 11 | Train      --> 10185/10185 batches | Elapsed time: 02:11:05 | Loss: 0.843 | Acc: 0.804


[INFO] Iteration 01 | Epoch 11 | Val        --> 348/348 batches | Elapsed time: 00:02:51 | Loss: 0.173 | Acc: 0.947


[INFO] Iteration 01 | Epoch 12 | Train      --> 10185/10185 batches | Elapsed time: 02:03:59 | Loss: 0.813 | Acc: 0.81


[INFO] Iteration 01 | Epoch 12 | Val        --> 348/348 batches | Elapsed time: 00:02:53 | Loss: 0.158 | Acc: 0.953


[INFO] Iteration 01 | Epoch 13 | Train      --> 10185/10185 batches | Elapsed time: 02:03:50 | Loss: 0.79 | Acc: 0.816


[INFO] Iteration 01 | Epoch 13 | Val        --> 348/348 batches | Elapsed time: 00:02:51 | Loss: 0.153 | Acc: 0.953


[INFO] Iteration 01 | Epoch 14 | Train      --> 10185/10185 batches | Elapsed time: 02:05:00 | Loss: 0.768 | Acc: 0.821


[INFO] Iteration 01 | Epoch 14 | Val        --> 348/348 batches | Elapsed time: 00:02:46 | Loss: 0.137 | Acc: 0.96


[INFO] Iteration 01 | Epoch 15 | Train      --> 10185/10185 batches | Elapsed time: 02:12:15 | Loss: 0.746 | Acc: 0.826


[INFO] Iteration 01 | Epoch 15 | Val        --> 348/348 batches | Elapsed time: 00:02:53 | Loss: 0.14 | Acc: 0.958


In [11]:
model_data = torch.load(f"{dataset_dir}/checkpoint.inc.lepidoptera.ta.i01.e14.pth", weights_only=False)
test_top_k(model_data, f"{dataset_dir}/test", 3, print_preds=False, print_top1_accuracy=True, print_no_match=True)
test_top_k(model_data, f"{dataset_dir}/test", 5, print_preds=False, print_top1_accuracy=False)
test_top_k(model_data, f"{dataset_dir}/test", 10, print_preds=False, print_top1_accuracy=False)

catopsilia-pomona        : belenois-aurota(0.439)  appias-libythea(0.178)  colias-erate(0.139)  
unidentified-butterfly-1 : neptis-nata(0.596)  neptis-clinia(0.304)  neptis-soma(0.038)  
unidentified-moth-2      : orgyia-postica-early(0.636)  asura-conferta(0.205)  nepita-conferta-early(0.065)  
unidentified-moth-3a     : pyralis-pictalis(0.423)  marasmia-poeyalis(0.247)  chlumetia-transversa(0.075)  
unidentified-moth-3b     : pyralis-manihotalis(0.646)  locastra-muscosalis(0.311)  syntypistis-pallidifascia(0.024)  
unidentified-moth-4a     : cleora-injectaria(0.486)  cleora-fraterna(0.319)  cleora-alienaria(0.106)  
unidentified-moth-4b     : cleora-injectaria(0.719)  boarmiini-genera-spp(0.084)  hypomecis-cineracea(0.040)  
unidentified-moth-5      : athetis-bremusa(0.460)  mythimna-separata(0.189)  spodoptera-mauritia(0.171)  
vindula-dejone           : cirrochroa-tyche(0.893)  vindula-erota(0.101)  cirrochroa-aoris(0.003)  
Top   1 accuracy: 38/49 -> 77.55%, genus matched: 46/49 -

In [12]:
model_data = torch.load(f"{dataset_dir}/checkpoint.inc.lepidoptera.ta.i01.e14.pth", weights_only=False)
test_top_k(model_data, f"{dataset_dir}/test2", 3, print_preds=False, print_top1_accuracy=True, print_no_match=False)
test_top_k(model_data, f"{dataset_dir}/test2", 5, print_preds=False, print_top1_accuracy=False)
test_top_k(model_data, f"{dataset_dir}/test2", 10, print_preds=False, print_top1_accuracy=False)

Top   1 accuracy: 242/308 -> 78.57%, genus matched: 282/308 -> 91.56%
Top   3 accuracy: 277/308 -> 89.94%, genus matched: 295/308 -> 95.78%
Top   5 accuracy: 288/308 -> 93.51%, genus matched: 297/308 -> 96.43%
Top  10 accuracy: 295/308 -> 95.78%, genus matched: 302/308 -> 98.05%


In [5]:
model_data = torch.load(f"{dataset_dir}/checkpoint.inc.lepidoptera.ta.i01.e15.pth", weights_only=False)
for epoch in range(5):
    model_data = run_epoch(model_data, output_path=f"{dataset_dir}/checkpoint.inc.lepidoptera.ta", robustness_lambda=0.1)

[INFO] Iteration 01 | Epoch 16 | Train      --> 10185/10185 batches | Elapsed time: 02:03:38 | Loss: 0.728 | Acc: 0.831


[INFO] Iteration 01 | Epoch 16 | Val        --> 348/348 batches | Elapsed time: 00:02:31 | Loss: 0.131 | Acc: 0.961


[INFO] Iteration 01 | Epoch 17 | Train      --> 10185/10185 batches | Elapsed time: 01:52:37 | Loss: 0.711 | Acc: 0.833


[INFO] Iteration 01 | Epoch 17 | Val        --> 348/348 batches | Elapsed time: 00:02:29 | Loss: 0.132 | Acc: 0.96


[INFO] Iteration 01 | Epoch 18 | Train      --> 10185/10185 batches | Elapsed time: 01:51:45 | Loss: 0.694 | Acc: 0.838


[INFO] Iteration 01 | Epoch 18 | Val        --> 348/348 batches | Elapsed time: 00:02:29 | Loss: 0.125 | Acc: 0.963


[INFO] Iteration 01 | Epoch 19 | Train      --> 10185/10185 batches | Elapsed time: 02:39:02 | Loss: 0.679 | Acc: 0.841


[INFO] Iteration 01 | Epoch 19 | Val        --> 348/348 batches | Elapsed time: 00:02:32 | Loss: 0.12 | Acc: 0.963


[INFO] Iteration 01 | Epoch 20 | Train      --> 10185/10185 batches | Elapsed time: 02:00:20 | Loss: 0.669 | Acc: 0.843


[INFO] Iteration 01 | Epoch 20 | Val        --> 348/348 batches | Elapsed time: 00:02:37 | Loss: 0.128 | Acc: 0.961


In [9]:
model_data = torch.load(f"{dataset_dir}/checkpoint.inc.lepidoptera.ta.i01.e19.pth", weights_only=False)
test_top_k(model_data, f"{dataset_dir}/test", 3, print_preds=False, print_top1_accuracy=True, print_no_match=True)
test_top_k(model_data, f"{dataset_dir}/test", 5, print_preds=False, print_top1_accuracy=False)
test_top_k(model_data, f"{dataset_dir}/test", 10, print_preds=False, print_top1_accuracy=False)

amyna-axis               : nolinae-genera-spp(0.596)  athetis-lineosa(0.178)  athetis-spp(0.094)  
unidentified-butterfly-1 : neptis-harita(0.645)  neptis-clinia(0.314)  neptis-neptis-spp(0.020)  
unidentified-moth-2      : orgyia-postica-early(0.813)  asura-conferta(0.107)  lymantria-ampla(0.056)  
unidentified-moth-3a     : pyralidae-genera-spp(0.496)  epipaschiinae-genera-spp(0.118)  locastra-muscosalis(0.085)  
unidentified-moth-3b     : pyralis-manihotalis(0.625)  locastra-muscosalis(0.198)  aporodes-floralis(0.131)  
unidentified-moth-4a     : cleora-injectaria(0.455)  cleora-alienaria(0.275)  cleora-spp(0.117)  
unidentified-moth-4b     : cleora-injectaria(0.979)  cleora-alienaria(0.007)  cleora-spp(0.005)  
unidentified-moth-5      : athetis-bremusa(0.351)  spodoptera-mauritia(0.260)  dichagyris-flammatra(0.158)  
vindula-dejone           : cirrochroa-tyche(0.947)  vindula-erota(0.046)  heliophorus-moorei(0.002)  
Top   1 accuracy: 39/49 -> 79.59%, genus matched: 47/49 -> 95.92

In [10]:
model_data = torch.load(f"{dataset_dir}/checkpoint.inc.lepidoptera.ta.i01.e19.pth", weights_only=False)
test_top_k(model_data, f"{dataset_dir}/test2", 3, print_preds=False, print_top1_accuracy=True, print_no_match=False)
test_top_k(model_data, f"{dataset_dir}/test2", 5, print_preds=False, print_top1_accuracy=False)
test_top_k(model_data, f"{dataset_dir}/test2", 10, print_preds=False, print_top1_accuracy=False)

Top   1 accuracy: 242/308 -> 78.57%, genus matched: 284/308 -> 92.21%
Top   3 accuracy: 281/308 -> 91.23%, genus matched: 299/308 -> 97.08%
Top   5 accuracy: 289/308 -> 93.83%, genus matched: 300/308 -> 97.40%
Top  10 accuracy: 295/308 -> 95.78%, genus matched: 302/308 -> 98.05%


## Iteration 2

- added distillation loss from teacher model (distill_lambda=1.0, temperature=2.0)

In [ ]:
model_data = torch.load(f"{dataset_dir}/checkpoint.inc.lepidoptera.ta.i01.e19.pth", weights_only=False)
model_data = init_iteration(model_data, f"{dataset_dir}/i01-train", f"{dataset_dir}/i01-val", lr=1e-4)
for epoch in range(5):
    model_data = run_epoch(model_data, output_path=f"{dataset_dir}/checkpoint.inc.lepidoptera.ta", robustness_lambda=0.1)

In [ ]:
model_data = torch.load(f"{dataset_dir}/checkpoint.inc.lepidoptera.ta.i02.e05.pth", weights_only=False)
test_top_k(model_data, f"{dataset_dir}/test", 3, print_preds=False, print_top1_accuracy=True, print_no_match=True)
test_top_k(model_data, f"{dataset_dir}/test", 5, print_preds=False, print_top1_accuracy=False)
test_top_k(model_data, f"{dataset_dir}/test", 10, print_preds=False, print_top1_accuracy=False)

In [ ]:
model_data = torch.load(f"{dataset_dir}/checkpoint.inc.lepidoptera.ta.i02.e05.pth", weights_only=False)
test_top_k(model_data, f"{dataset_dir}/test2", 3, print_preds=False, print_top1_accuracy=True, print_no_match=False)
test_top_k(model_data, f"{dataset_dir}/test2", 5, print_preds=False, print_top1_accuracy=False)
test_top_k(model_data, f"{dataset_dir}/test2", 10, print_preds=False, print_top1_accuracy=False)